### ISSUE TITLE
Wrapping bounds in tt.all changes log likelihood

### ISSUE THREAD

**Issue by AustinRochford** [pymc-devs/pymc#1449]
Here I have two implementations of a multinomial random variable, which differ only in the shape of the conditions passed to `bound`.  Models that only differ in this way produce different log likelihoods for reasons that are not clear to me.

``` python
import numpy as np
import pymc3 as pm
from pymc3.distributions import Discrete
from pymc3.distributions.dist_math import bound, factln
from theano import tensor as tt
```

``` python
x = np.array([1, 5])
n = x.sum()
```

``` python
class MultinomialA(Discrete):
    def __init__(self, n, p, *args, **kwargs):
        super(MultinomialA, self).__init__(*args, **kwargs)

        self.n = n
        self.p = p

    def logp(self, value):
        n = self.n
        p = self.p

        return bound(factln(n) - factln(value).sum() + (value * tt.log(p)).sum(),
                     value >= 0,
                     0 <= p, p <= 1,
                     tt.isclose(p.sum(), 1))
```

``` python
with pm.Model() as modelA:
    p = pm.Dirichlet('p', np.ones(2))
    x_obs = MultinomialA('x', n, p, observed=x)
```

```
Applied stickbreaking-transform to p and added transformed p_stickbreaking_ to model.
```

``` python
modelA.logp({'p_stickbreaking_': [0]})
```

```
array(-6.120541589383123)
```

``` python
class MultinomialB(Discrete):
    def __init__(self, n, p, *args, **kwargs):
        super(MultinomialB, self).__init__(*args, **kwargs)

        self.n = n
        self.p = p

    def logp(self, value):
        n = self.n
        p = self.p

        return bound(factln(n) - factln(value).sum() + (value * tt.log(p)).sum(),
                     tt.all(value >= 0),
                     tt.all(0 <= p), tt.all(p <= 1),
                     tt.isclose(p.sum(), 1))
```

``` python
with pm.Model() as modelB:
    p = pm.Dirichlet('p', np.ones(2))
    x_obs = MultinomialB('x', n, p, observed=x)
```

```
Applied stickbreaking-transform to p and added transformed p_stickbreaking_ to model.
```

``` python
modelB.logp({'p_stickbreaking_': [0]})
```

```
array(-3.753417975251507)
```

I think this may be a bug, but I may also be misunderstanding the expected behavior of `bound`.

**Comment by twiecki** [comment]
That is super odd indeed. I expect the bounds to not affect the logp if they are not triggered. Can you post the NB to make it easier to work with?

**Comment by AustinRochford** [comment]
@twiecki https://gist.github.com/AustinRochford/5b63cd542b9b727dc41c8c2642da30ab

**Comment by AustinRochford** [comment]
One guess: the [current implementation](https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/dist_math.py#L31-L35) of `alltrue` is doing some broadcasting that we're not expecting.

**Comment by AustinRochford** [comment]
That is the problem; I will submit a fix tonight after work.

**Comment by ColCarroll** [comment]
@AustinRochford : check out #394 and https://github.com/Theano/Theano/issues/1615 for previous discussion of `alltrue`.

**Comment by AustinRochford** [comment]
I have an idea for how to fix this, but I can't get the tests to pass on master on my macbook, so I'm at a bit of a loss at the moment on how to gain confidence in my changes.

```
> nosetests -xv pymc3.tests.test_diagnostics
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 1216.87it/s]
ok
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 612.71it/s]
ok
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 610.95it/s]
ok
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 608.46it/s]
ok
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 612.48it/s]
ok
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 446.93it/s]
ok
Confirm Geweke diagnostic is larger than 1 for a small number of samples. ... ok
Confirm Geweke diagnostic is smaller than 1 for a reasonable number of samples. ... ok
Confirm Gelman-Rubin statistic is far from 1 for a small number of samples. ... FAIL

======================================================================
FAIL: Confirm Gelman-Rubin statistic is far from 1 for a small number of samples.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/austin/Documents/Projects/pymc3/pymc3/tests/test_diagnostics.py", line 41, in test_bad
    self.good_ratio for r in rhat.values()))
AssertionError: True is not false
-------------------- >> begin captured logging << --------------------
pymc3: INFO: Applied log-transform to early_mean and added transformed early_mean_log_ to model.
pymc3: INFO: Applied log-transform to late_mean and added transformed late_mean_log_ to model.
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 9 tests in 17.664s

FAILED (failures=1)
```

**Comment by ColCarroll** [comment]
That should be a seeded test, but isn't -- is it failing consistently?

**Comment by AustinRochford** [comment]
@ColCarroll I have run it ~10 times, and it has failed every time.  It is also the first among many failures/errors. I have taken to stopping the test run at the first failure due to the amount of time the full test suite takes to run.

**Comment by ColCarroll** [comment]
That's strange -- I just opened a PR to make the tests reproducible, but it
should _usually_ pass, at least, so I'm not sure this actually fixes your
problem.

**Comment by AustinRochford** [comment]
Can't hurt to try again once your PR is merged.  Otherwise, I'll try again tomorrow.

**Comment by AustinRochford** [comment]
It seems that while I can't get them to pass on my Mac for Python 2.7, I can get them to pass for Python 3.5.  This is enough for me to go off of locally, and rely on Travis for the other versions.  Will put up a PR over the weekend.

**Comment by twiecki** [comment]
OK, seems like an odd issue, hopefully travis agrees.

**Comment by twiecki** [comment]
Closed by #1452. Thanks @AustinRochford!
